### ISSUE TITLE
Adding eigenvalue and eigenvector computations

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#2618]
NumPy's [`linalg`]( https://docs.scipy.org/doc/numpy/reference/routines.linalg.html ) includes a variety of eigenvalue and eigenvector functions (e.g. [`eig`, `eigh`, `eigvals`, `eigvalsh`]( https://docs.scipy.org/doc/numpy/reference/routines.linalg.html#matrix-eigenvalues ). Would be nice to include these in Dask Array's `linalg` as well.

**Comment by mrocklin** [comment]
We would probably have to do a bit of scholarly research to find out the right parallel algorithms for these.

**Comment by tmoerman** [comment]
+1 for this feature

**Comment by TomAugspurger** [comment]
@tmoerman any interest in working on this? ;)

I'll have a need for this in 1-2 months, so I'll probably start on it in a couple weeks.

**Comment by mrocklin** [comment]
@TomAugspurger I'm curious what your need is

**Comment by jmswaney** [comment]
+1 for this. My use case is in feature extraction for large 3D images that don't fit into memory. At each voxel, however, I have a small 3x3 symmetric matrix that I'd like to find the eigenvalues for. Numpy makes this easy for smaller volumes since it just computes the eigenvalues along the last two dimensions.

**Comment by shoyer** [comment]
If you only need eigenvectorsfor a long set of 3x3 matrices, this should be straightforward to parallelize with with `atop`.

**Comment by ShivangiM** [comment]
Any progress in this? Also, any example to use atop to achieve the task of finding eigenvector and eigenvalue?

**Comment by tasansal** [comment]
I accomplished this via apply_gufunc from dask. I have a gradient structure tensor at a voxel in a 3D volume and it is shaped (Nx, Ny, Nz, 3, 3) where the last 2 dimensions are the 3x3 matrix I want to run eigh on.

I am not sure if this is the best way to do it, but the results look correct and my code runs faster than just eigh.. Haven't done any real profiling yet.

i,j are the principal dimensions (in this case it is the last 2).

w, v = apply_gufunc(eigh, '(i,j)->(i),(i,j)', gst)

Maybe dask experts will chime in and suggest a better way.

**Comment by mrocklin** [comment]
Nice example @tasansal !

In my ideal world, `np.linalg.eigh` would be a gufunc itself, so that you could just call it directly on `gst` and have it do the right thing.  

cc @seberg from Numpy

**Comment by seberg** [comment]
I agree, the main issue is probably that we would need to find a solution for custom keyword argument handling, which adds some new complexity to ufuncs and is not specified in `__array_ufunc__` as of now.

The linalg functions could probably rely more on the (currently not public) ufuncs inside, and e.g. try to avoid converting to `ndarray`, although I am not sure that helps (I expect they currently mainly converting to `ndarray` for `matrix` support...).

**Comment by tasansal** [comment]
@mrocklin @seberg,

I actually did some manual profiling. Here is my dask array.
dask.array<rechunk-merge, shape=(12, 454, 3701, 3, 3), dtype=float64, chunksize=(4, 227, 2051, 3, 3), chunktype=numpy.ndarray>

As you can see chunks are about 127MB.  Number of chunks is about 12. So if I was lucky and performance was linear it would speed up by 12x.

The vanilla numpy eigh on it takes 74 secs (uses 1/16 corse = 10% task manager).
The dask version takes 62 secs (utilizes 88% of 16 cores)

It doesn't seem to speed up that much. However, while I was testing, on different array sizes the performance improves quite significantly. I can't replicate it again at the moment. I think it has something to do with chunk sizes. 

Just wanted to give more insight... Currently both calls below work with same performance.
w, v = apply_gufunc(eigh, '(m,m)->(m),(m,m)', gst)
OR
w, v = np.linalg._umath_linalg.eigh_lo(gst, output_dtypes=('float', 'float')) (Saw on dask docs)

**Comment by seberg** [comment]
I am not sure how much work `eigh_lo` has to do, for those very small matrices – `3x3` in the last dimensions – as here (especially on such a huge datatset), on a single machine the speed is possibly bound only by memory/io speed.

I assume in your last call you mean `eigh_lo(gst, signature=(None, "float", "float"))` or is that a Dask thing that somehow manages to get through `__array_ufunc__`?!

**Comment by tasansal** [comment]
@seberg Regarding the second paragraph: it works with both signatures. Not sure how Dask handles it. @mrocklin any idea?

**Comment by seberg** [comment]
No, it is valid API for dask, and NumPy currently lets it pass. I am pretty sure that is not intentional and it might make cleaning up NumPy code harder, but that does not really matter here. Maybe we can even just adopt the additional `output_dtypes` in numpy (although that assumes this is the only important one being added).

**Comment by briley-lewis** [comment]
Just another +1 for this feature request! I need to do an eigendecomposition on a particularly large array (327,680 x 327,680) and am currently trying dask's svd, but a function similar to numpy's eigh would be great

**Comment by acs14007** [comment]
> In my ideal world, `np.linalg.eigh` would be a gufunc itself, so that you could just call it directly on `gst` and have it do the right thing.
